File: oma/core/edge.py
"""
OMA Edge Handlers -- graceful degradation at the limits.

Two edge cases, two strategies:

1. NEAR OUTAGE (completion almost done, tokens running low):
   Summarize with remaining tokens. Pack the handoff note with
   everything the next worker needs to immediately pick up.

2. OUTAGE BEFORE COMPLETION (unexpected cutoff):
   The handoff was already being maintained incrementally,
   so the last checkpoint is the recovery point. The next worker
   gets a note pointing to the fork and patches to apply.
"""

import json
import time
from dataclasses import dataclass, field
from typing import Any, Optional


@dataclass
class HandoffNote:
    """
    Everything the next worker needs to resume.
    This is what gets written to disk/memory before outage.
    """
    task_id: str
    objective: str
    status: str                              # near_outage | outage | parked
    reason: str
    timestamp: float = field(default_factory=time.time)

    # what was accomplished
    completed_steps: list = field(default_factory=list)
    artifacts_produced: dict = field(default_factory=dict)
    confidence_so_far: float = 0.0

    # what remains
    remaining_steps: list = field(default_factory=list)
    blockers: list = field(default_factory=list)

    # context the next worker needs immediately
    critical_context: str = ""               # the "grab this first" brief
    criteria: dict = field(default_factory=dict)
    provider_status: dict = field(default_factory=dict)

    # recovery instructions
    patches_to_apply: list = field(default_factory=list)
    binary_target: str = ""                  # e.g., "macos arm64"
    priority_order: list = field(default_factory=list)  # features by importance

    def to_prompt(self) -> str:
        """
        Render as a prompt the next worker can consume directly.
        This is the handoff document.
        """
        sections = []

        sections.append(f"=== OMA HANDOFF NOTE ===")
        sections.append(f"task: {self.task_id}")
        sections.append(f"objective: {self.objective}")
        sections.append(f"status: {self.status} ({self.reason})")
        sections.append(f"confidence: {self.confidence_so_far:.0%}")
        sections.append(f"time: {time.strftime('%Y-%m-%d %H:%M:%S UTC', time.gmtime(self.timestamp))}")

        if self.critical_context:
            sections.append(f"\n--- GRAB THIS FIRST ---")
            sections.append(self.critical_context)

        if self.completed_steps:
            sections.append(f"\n--- COMPLETED ---")
            for i, step in enumerate(self.completed_steps, 1):
                sections.append(f"  {i}. {step}")

        if self.remaining_steps:
            sections.append(f"\n--- REMAINING (do these) ---")
            for i, step in enumerate(self.remaining_steps, 1):
                sections.append(f"  {i}. {step}")

        if self.priority_order:
            sections.append(f"\n--- PRIORITY ORDER (most important first) ---")
            for i, feat in enumerate(self.priority_order, 1):
                sections.append(f"  {i}. {feat}")

        if self.patches_to_apply:
            sections.append(f"\n--- PATCHES TO APPLY ---")
            for patch in self.patches_to_apply:
                sections.append(f"  - {patch}")

        if self.binary_target:
            sections.append(f"\n--- BUILD TARGET ---")
            sections.append(f"  {self.binary_target}")

        if self.blockers:
            sections.append(f"\n--- BLOCKERS ---")
            for b in self.blockers:
                sections.append(f"  ! {b}")

        if self.artifacts_produced:
            sections.append(f"\n--- ARTIFACTS ---")
            for name, path in self.artifacts_produced.items():
                sections.append(f"  {name}: {path}")

        if self.criteria:
            sections.append(f"\n--- SUCCESS CRITERIA ---")
            sections.append(f"  {json.dumps(self.criteria, indent=2, default=str)}")

        if self.provider_status:
            sections.append(f"\n--- PROVIDER STATUS ---")
            for name, status in self.provider_status.items():
                sections.append(f"  {name}: {status}")

        sections.append(f"\n=== END HANDOFF ===")
        return "\n".join(sections)

    def to_dict(self) -> dict:
        return {
            "task_id": self.task_id,
            "objective": self.objective,
            "status": self.status,
            "reason": self.reason,
            "timestamp": self.timestamp,
            "completed_steps": self.completed_steps,
            "remaining_steps": self.remaining_steps,
            "priority_order": self.priority_order,
            "confidence": self.confidence_so_far,
            "artifacts": self.artifacts_produced,
            "criteria": self.criteria,
            "patches": self.patches_to_apply,
            "binary_target": self.binary_target,
            "critical_context": self.critical_context,
        }


def near_outage_handler(
    task_state,
    working_memory,
    remaining_features: list,
    patches: list = None,
) -> HandoffNote:
    """
    Called when tokens are running low but we can still write.
    Produces the most useful handoff possible with remaining budget.
    """
    completed = [
        f"{step}: conf={conf:.2f}"
        for step, result, conf in task_state.progress
        if conf > 0.3
    ]

    remaining = [
        f"{feat}" for feat in remaining_features
    ]

    return HandoffNote(
        task_id=task_state.task_id,
        objective=task_state.objective,
        status="near_outage",
        reason=f"tokens {task_state.tokens_used}/{task_state.tokens_budget}",
        completed_steps=completed,
        remaining_steps=remaining,
        confidence_so_far=task_state.confidence,
        artifacts_produced=dict(task_state.artifacts),
        criteria=task_state.criteria,
        patches_to_apply=patches or [],
        priority_order=remaining_features,
        critical_context=working_memory.summarize(max_tokens=2000),
    )


def outage_recovery_prompt(
    handoff_note: HandoffNote,
    fork_repo: str = "",
    target_platform: str = "macos",
) -> str:
    """
    Generate the prompt for the next worker after an outage.

    This is the "note to next worker" that includes everything
    needed to resume, including fork references and build targets.
    """
    base = handoff_note.to_prompt()

    recovery = [
        f"\n=== RECOVERY INSTRUCTIONS ===",
        f"From a fork of open-multi-agent, consider these patches",
        f"and create a working binary for {target_platform} on this computer",
        f"in order to complete the functionalities listed from most",
        f"important to least, checking off as many as possible.",
    ]

    if fork_repo:
        recovery.append(f"\nFork: {fork_repo}")

    recovery.append(f"\nFunctionalities (priority order):")
    for i, feat in enumerate(handoff_note.priority_order, 1):
        recovery.append(f"  [ ] {i}. {feat}")

    return base + "\n".join(recovery)

File: oma-ts/src/core/edge.ts
/**
 * OMA Edge Handlers - graceful degradation at the limits.
 *
 * Two edge cases, two strategies:
 *
 * 1. NEAR OUTAGE (completion almost done, tokens running low):
 *    Summarize with remaining tokens. Pack the handoff note with
 *    everything the next worker needs to immediately pick up.
 *
 * 2. OUTAGE BEFORE COMPLETION (unexpected cutoff):
 *    The handoff was already being maintained incrementally,
 *    so the last checkpoint is the recovery point. The next worker
 *    gets a note pointing to the fork and patches to apply.
 */

import type { TaskState } from './loop.js';
import type { WorkingMemory } from '../automation/memory.js';

export interface HandoffNoteInit {
  task_id: string;
  objective: string;
  status: string;
  reason: string;
  timestamp?: number;
  completed_steps?: string[];
  artifacts_produced?: Record<string, string>;
  confidence_so_far?: number;
  remaining_steps?: string[];
  blockers?: string[];
  critical_context?: string;
  criteria?: Record<string, unknown>;
  provider_status?: Record<string, string>;
  patches_to_apply?: string[];
  binary_target?: string;
  priority_order?: string[];
}

export class HandoffNote {
  task_id: string;
  objective: string;
  status: string;
  reason: string;
  timestamp: number;
  completed_steps: string[];
  artifacts_produced: Record<string, string>;
  confidence_so_far: number;
  remaining_steps: string[];
  blockers: string[];
  critical_context: string;
  criteria: Record<string, unknown>;
  provider_status: Record<string, string>;
  patches_to_apply: string[];
  binary_target: string;
  priority_order: string[];

  constructor(init: HandoffNoteInit) {
    this.task_id = init.task_id;
    this.objective = init.objective;
    this.status = init.status;
    this.reason = init.reason;
    this.timestamp = init.timestamp ?? Date.now() / 1000;
    this.completed_steps = init.completed_steps ?? [];
    this.artifacts_produced = init.artifacts_produced ?? {};
    this.confidence_so_far = init.confidence_so_far ?? 0;
    this.remaining_steps = init.remaining_steps ?? [];
    this.blockers = init.blockers ?? [];
    this.critical_context = init.critical_context ?? '';
    this.criteria = init.criteria ?? {};
    this.provider_status = init.provider_status ?? {};
    this.patches_to_apply = init.patches_to_apply ?? [];
    this.binary_target = init.binary_target ?? '';
    this.priority_order = init.priority_order ?? [];
  }

  toPrompt(): string {
    const sections: string[] = [];
    const ts = new Date(this.timestamp * 1000).toISOString().replace('T', ' ').replace(/\.\d+Z/, ' UTC');

    sections.push('=== OMA HANDOFF NOTE ===');
    sections.push(`task: ${this.task_id}`);
    sections.push(`objective: ${this.objective}`);
    sections.push(`status: ${this.status} (${this.reason})`);
    sections.push(`confidence: ${Math.round(this.confidence_so_far * 100)}%`);
    sections.push(`time: ${ts}`);

    if (this.critical_context) {
      sections.push('\n--- GRAB THIS FIRST ---');
      sections.push(this.critical_context);
    }

    if (this.completed_steps.length > 0) {
      sections.push('\n--- COMPLETED ---');
      this.completed_steps.forEach((step, i) => sections.push(`  ${i + 1}. ${step}`));
    }

    if (this.remaining_steps.length > 0) {
      sections.push('\n--- REMAINING (do these) ---');
      this.remaining_steps.forEach((step, i) => sections.push(`  ${i + 1}. ${step}`));
    }

    if (this.priority_order.length > 0) {
      sections.push('\n--- PRIORITY ORDER (most important first) ---');
      this.priority_order.forEach((feat, i) => sections.push(`  ${i + 1}. ${feat}`));
    }

    if (this.patches_to_apply.length > 0) {
      sections.push('\n--- PATCHES TO APPLY ---');
      this.patches_to_apply.forEach(p => sections.push(`  - ${p}`));
    }

    if (this.binary_target) {
      sections.push('\n--- BUILD TARGET ---');
      sections.push(`  ${this.binary_target}`);
    }

    if (this.blockers.length > 0) {
      sections.push('\n--- BLOCKERS ---');
      this.blockers.forEach(b => sections.push(`  ! ${b}`));
    }

    if (Object.keys(this.artifacts_produced).length > 0) {
      sections.push('\n--- ARTIFACTS ---');
      for (const [name, path] of Object.entries(this.artifacts_produced)) {
        sections.push(`  ${name}: ${path}`);
      }
    }

    if (Object.keys(this.criteria).length > 0) {
      sections.push('\n--- SUCCESS CRITERIA ---');
      sections.push(`  ${JSON.stringify(this.criteria, null, 2)}`);
    }

    if (Object.keys(this.provider_status).length > 0) {
      sections.push('\n--- PROVIDER STATUS ---');
      for (const [name, status] of Object.entries(this.provider_status)) {
        sections.push(`  ${name}: ${status}`);
      }
    }

    sections.push('\n=== END HANDOFF ===');
    return sections.join('\n');
  }

  toDict(): Record<string, unknown> {
    return {
      task_id: this.task_id,
      objective: this.objective,
      status: this.status,
      reason: this.reason,
      timestamp: this.timestamp,
      completed_steps: this.completed_steps,
      remaining_steps: this.remaining_steps,
      priority_order: this.priority_order,
      confidence: this.confidence_so_far,
      artifacts: this.artifacts_produced,
      criteria: this.criteria,
      patches: this.patches_to_apply,
      binary_target: this.binary_target,
      critical_context: this.critical_context,
    };
  }
}

export function nearOutageHandler(
  taskState: TaskState,
  workingMemory: WorkingMemory,
  remainingFeatures: string[],
  patches: string[] = [],
): HandoffNote {
  const completed = taskState.progress
    .filter(([, , conf]) => conf > 0.3)
    .map(([step, , conf]) => `${step}: conf=${conf.toFixed(2)}`);

  return new HandoffNote({
    task_id: taskState.task_id,
    objective: taskState.objective,
    status: 'near_outage',
    reason: `tokens ${taskState.tokens_used}/${taskState.tokens_budget}`,
    completed_steps: completed,
    remaining_steps: remainingFeatures,
    confidence_so_far: taskState.confidence,
    artifacts_produced: { ...taskState.artifacts },
    criteria: taskState.criteria,
    patches_to_apply: patches,
    priority_order: remainingFeatures,
    critical_context: workingMemory.summarize(2000),
  });
}

export function outageRecoveryPrompt(
  handoffNote: HandoffNote,
  forkRepo = '',
  targetPlatform = 'macos',
): string {
  const base = handoffNote.toPrompt();

  const recovery = [
    '\n=== RECOVERY INSTRUCTIONS ===',
    'From a fork of open-multi-agent, consider these patches',
    `and create a working binary for ${targetPlatform} on this computer`,
    'in order to complete the functionalities listed from most',
    'important to least, checking off as many as possible.',
  ];

  if (forkRepo) recovery.push(`\nFork: ${forkRepo}`);

  recovery.push('\nFunctionalities (priority order):');
  handoffNote.priority_order.forEach((feat, i) => {
    recovery.push(`  [ ] ${i + 1}. ${feat}`);
  });

  return base + recovery.join('\n');
}
